azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,325 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import logging
+import uuid
+from threading import Timer
+from transitions import Machine
+from ..transport import constant
+
+from .action import SubscribeAction, PublishAction, UnsubscribeAction
+from .request_response_provider import RequestResponseMachine
+
+
+logger = logging.getLogger(__name__)
+
+
+class PollingMachine(object):
+    """
+    Class that is responsible for sending the initial registration request and polling the
+    registration process for constant updates.
+    """
+
+    def __init__(self, state_based_provider):
+        """
+        :param state_based_provider: The state machine based provider.
+        """
+        self._polling_timer = None
+        self._query_timer = None
+
+        self._register_callback = None
+        self._cancel_callback = None
+
+        self.on_connected = None
+        self.on_disconnected = None
+        self.on_registration_status_update = None
+
+        self._operations = {}
+
+        self._request_response_provider = RequestResponseMachine(state_based_provider)
+
+        self._request_response_provider.on_response_received = self._handle_response_received
+
+        states = [
+            "disconnected",
+            "initializing",
+            "registering",
+            "waiting_to_poll",
+            "polling",
+            "completed",
+            "error",
+            "cancelling",
+        ]
+
+        transitions = [
```

Comment:
I noticed some things are undefined in the state machine:

* Error trigger during initializing
* Cancellation during initializing
* Register triggered during waiting_to_poll, polling, or initializing?

* Once Completed or Cancelling state reached... nothing more can ever happen?

Reply:
• Error trigger during initializing - Not sure how will this occur other than receiving exception from connect or subscribe and dealing with it by calling trigger_error.
• Cancellation during initializing - This can be included but its right after initializing it goes to registering. I will include it for now but I am unable to simulate the behavior.
• Register triggered during waiting_to_poll, polling, or initializing? I don't think this can occur as all these states only happen after user calls register , so it can't be triggered when in these states.
• Once Completed or Cancelling state reached... nothing more can ever happen? Yeah that is how I designed it.


---
In reply to: [277804489](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277804489) [](ancestors = 277804489)